add support for rapidfuzz

maxbachmann opened the pull request:
The implementation used by rapidfuzz has the following algorithms
- Jaro/JaroWinkler (fastest by a large margin)
- Hamming (slightly slower than python-Levenshtein)
- Levenshtein (similar fast to python-Levenshtein for very short strings and fastest for longer strings)

Additionally it supports any sequence of hashable types (e.g. lists of strings) and not only text

Here is the benchmark result:
```
# Faster than textdistance:

| algorithm          | library                 | function                     |        time |
|--------------------+-------------------------+------------------------------+-------------|
| DamerauLevenshtein | jellyfish               | damerau_levenshtein_distance | 0.0181046   |
| DamerauLevenshtein | pyxdameraulevenshtein   | damerau_levenshtein_distance | 0.030925    |
| Hamming            | Levenshtein             | hamming                      | 0.000351586 |
| Hamming            | rapidfuzz.string_metric | hamming                      | 0.00040442  |
| Hamming            | jellyfish               | hamming_distance             | 0.0143502   |
| Jaro               | rapidfuzz.string_metric | jaro_similarity              | 0.000749048 |
| Jaro               | jellyfish               | jaro_similarity              | 0.0152322   |
| JaroWinkler        | rapidfuzz.string_metric | jaro_winkler_similarity      | 0.000776006 |
| JaroWinkler        | jellyfish               | jaro_winkler_similarity      | 0.0157833   |
| Levenshtein        | rapidfuzz.string_metric | levenshtein                  | 0.0010058   |
| Levenshtein        | Levenshtein             | distance                     | 0.00103176  |
| Levenshtein        | jellyfish               | levenshtein_distance         | 0.0147382   |
| Levenshtein        | pylev                   | levenshtein                  | 0.14116     |
Total: 13 libs.
```

and the benchmark results when adding slightly longer strings:
```python
STMT = """
func('text', 'test')
func('qwer', 'asdf')
func('a' * 15, 'b' * 15)
func('a' * 30, 'b' * 30)
"""
```
```
# Faster than textdistance:

| algorithm          | library                 | function                     |        time |
|--------------------+-------------------------+------------------------------+-------------|
| DamerauLevenshtein | jellyfish               | damerau_levenshtein_distance | 0.0323887   |
| DamerauLevenshtein | pyxdameraulevenshtein   | damerau_levenshtein_distance | 0.143235    |
| Hamming            | Levenshtein             | hamming                      | 0.000489837 |
| Hamming            | rapidfuzz.string_metric | hamming                      | 0.000517879 |
| Hamming            | jellyfish               | hamming_distance             | 0.0182341   |
| Jaro               | rapidfuzz.string_metric | jaro_similarity              | 0.00111363  |
| Jaro               | jellyfish               | jaro_similarity              | 0.0201971   |
| JaroWinkler        | rapidfuzz.string_metric | jaro_winkler_similarity      | 0.00105238  |
| JaroWinkler        | jellyfish               | jaro_winkler_similarity      | 0.0206678   |
| Levenshtein        | rapidfuzz.string_metric | levenshtein                  | 0.00138601  |
| Levenshtein        | Levenshtein             | distance                     | 0.0034889   |
| Levenshtein        | jellyfish               | levenshtein_distance         | 0.0232467   |
| Levenshtein        | pylev                   | levenshtein                  | 0.599603    |
Total: 13 libs.
```

maxbachmann commented on textdistance/libraries.py:
```diff
@@ -97,6 +97,12 @@ def get_function(self):
                 obj = getattr(module, self.func_name)().dist_abs
+            if self.module_name == 'rapidfuzz.string_metric':
+                if self.func_name in {'jaro_similarity', 'jaro_winkler_similarity'}:
+                    func = getattr(module, self.func_name)
+                    obj = lambda *args: func(*args) / 100
```
For similarities RapidFuzz uses a range 0.0-100.0 instead of 0.0-1.0, so divide by 100

maxbachmann commented on textdistance/libraries.py:
```diff
@@ -97,6 +97,16 @@ def get_function(self):
                 # called dist_abs() (whereas dist() gives a normalised distance)
                 obj = getattr(module, self.func_name)().dist_abs
+            if self.module_name == 'rapidfuzz.string_metric':
+                if self.func_name in {'jaro_similarity', 'jaro_winkler_similarity'}:
+                    module_func = getattr(module, self.func_name)
+
+                    def func(s1, s2):
+                        return module_func(s1, s2) / 100
+
+                    obj = func
```
Hm not sure pep8 made this a lot nicer than
```
module_func = getattr(module, self.func_name)
obj = lambda s1, s2: module_func(s1, s2) / 100
```

maxbachmann commented:
@orsinium I am not quite sure how to fix the build errors, since I am not very familiar with drone CI + alpine linux. Currently numpy + rapidfuzz do not have musllinux wheels, so it needs to build everything from source. Is there any specific reason why the CI uses alpine linux? Pretty much no project has wheels for musllinux so far.

orsinium commented:
Thank you! Looks cool. Please, resolve CI, and I'll merge it. No, there is no particular reason to use alpine except that it worked well before. You can migrate it to anything else, buster should do. I think it should be enough. Give it a try and if you can't make it work, just ping me, I'll help.

maxbachmann commented:
@orsinium This will likely take a bit longer, since I am currently working on v2.0.0 of RapidFuzz, which will simplify this PR.

I have one question regarding textdistance:
I generally like the way you classify metrics into different categories like e.g.edit based distances and plan to do something similar when adding more metrics to RapidFuzz. However I was unsure when a metric is `sequence_based`. I would have personally placed LCSSeq in `edit_based`, since it is a metric which only uses Insertions/Deletions but no Substitutions.

orsinium commented:
If I remember correctly, sequence_based are ones that find common subsequences in both given strings.

maxbachmann commented:
@orsinium I finally had time for this. I updated rapidfuzz, which simplifies the addition. In addition I updated the ci to debian. As a side effect this appears to decrease the CI runtime significantly.

I do not understand the test failures. From reading it sound like the following occurs:
```
rapidfuzz.distance.Jaro.similarity('0', '0') -> 0.0
rapidfuzz.distance.JaroWinkler.similarity('0', '0') -> 0.0
rapidfuzz.distance.Jaro.similarity([], []) -> 0.0
rapidfuzz.distance.JaroWinkler.similarity([], []) -> 0.0
```

I can not reproduce the first two issues:
```
>>> from rapidfuzz.distance import Jaro, JaroWinkler
>>> Jaro.similarity('0', '0')
1.0
>>> JaroWinkler.similarity('0', '0')
1.0
```

the other one comes down to the question, whether the result of matching two empty sequences should be a perfect match or no match.
```
>>> textdistance.algorithms.Jaro().similarity('', '')
1
>>> textdistance.algorithms.Jaro().similarity([], [])
1
>>> rapidfuzz.distance.Jaro.similarity('', '')
0.0
>>> rapidfuzz.distance.Jaro.similarity([], [])
0.0
>>> jellyfish.jaro_similarity('', '')
0.0
>>> Levenshtein.jaro_winkler('', '')
1.0
```

maxbachmann commented:
As far as I can see the behavior for two empty sequences does not matter, since it will not be called with two empty sequences anyways:
```python
        if not all(sequences):
            return self.maximum(*sequences)
        # try get answer from external libs
        answer = self.external_answer(*sequences)
        if answer is not None:
            return answer
```

orsinium commented:
The point of the two failed tests is to ensure that doesn't matter if you use an external or internal implementation of an algorithm, the result will be the same in both cases. If there is a mismatch, we need to address it on either side.

> I can not reproduce the first two issues

The test `test_qval` converts the input sequence into q-grams before running the external function. So, I guess, the input that rapidfuzz gets is something like `[('0',)]`, not just `'0'`. You can run pytest with `--pdb` to see for sure.

orsinium commented:
If you have issues only with Jaro and Jaro-Winkler, you can disable rapidfuzz for them for now and merge only integration for Hamming and Levenshtein. You've done quite some work, and having something merged is better than nothing :)

orsinium commented on textdistance/libraries.json:
```diff
@@ -19,39 +15,39 @@
+      "distance"
     ],
     [
-      "abydos.distance",
```
Did you remove abydos from the list? Why? Is this file autogenerated? I don't remember 🤔  Why is it not in the list anymore, do you have an idea?

maxbachmann replied:
I think this was generated and I did not have abydos on my machine at the time (abydos still fails on python3.10 for me). Manually added them back

maxbachmann commented on tests/test_external.py:
```diff
@@ -62,7 +62,7 @@ def test_qval(left, right, alg):
             s1, s2 = lib.prepare(s1, s2)
             ext_result = external_func(s1, s2)
-        assert isclose(int_result, ext_result), str(lib)
+            assert isclose(int_result, ext_result), str(lib)
```
this should be a bug in the tests

maxbachmann commented:
@orsinium fixed the remaining issues. Somehow my last changes caused the flake8 test to fail, which worked yesterday. Not quite sure what this is about.

maxbachmann commented:
> Somehow my last changes caused the flake8 test to fail, which worked yesterday. Not quite sure what this is about.

worked after re-triggering the CI.

rebased everything on master and should be ready to merge.

orsinium commented on textdistance/libraries.py:
```diff
@@ -97,6 +97,18 @@ def get_function(self):
                 # object constructor - the distance metric method is
                 # called dist_abs() (whereas dist() gives a normalised distance)
                 obj = getattr(module, self.func_name)().dist_abs
+            elif self.module_name in {'rapidfuzz.distance.Jaro', 'rapidfuzz.distance.JaroWinkler'}:
```
Interesting 🤔 So, rapidfuzz considers two empty strings to have 0 similarity? Why? An empty string is equal to another empty string.

orsinium replied:
Also, it does it only for Jaro but not for Hamming, if I understand correctly. Very interesting. Can we fix it upstream? Or there is anything in the original algorithm that says otherwise?

maxbachmann replied:
Similar to `textdistance` I have the following algorithms
```
distance
similarity
normalized_similarity
normalized_distance
```

for two empty strings I handle this in the following way:
```
distance -> 0
similarity -> max - distance = 0 - 0
normalized_similarity -> 1.0 - normalized_distance -> 1.0
normalized_distance -> max / dist if max else 0 -> 0
```

orsinium commented:
Now that should do. I've patched tests to skip quick results since there are calculated before calling any external libs anyway.

orsinium commented:
Thank you!
